### PR TITLE
sql: support USING on implicitly coercible data types

### LINKED ISSUE #5276: USING/Natural joins should implicitly cast join key columns
```
arjun@altaria  ~/tmp  psql -d materialize -p 6875 -h localhost               
materialize=> create table intraday (listing int8);
CREATE TABLE
materialize=> create table fix (listing int4);l
CREATE TABLE
materialize=> create materialized view position as select * from intraday full outer join fix using (listing);
ERROR:  Int64 and Int32 are not comparable (in NATURAL/USING join on listing)
materialize=> select mz_version();
       mz_version       
------------------------
 v0.6.1-dev (1a82bf629)
```

### PR DESCRIPTION
**Pull request by sploiselle**
Fixes #5276

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5313)

### CODE REVIEW

File: test/testdrive/joins.td [hunk 1]
```diff
@@ -239,3 +239,41 @@ Unable to automatically determine a timestamp for your query; this can happen if
 
 ! SELECT * FROM names FULL OUTER JOIN mods_unmat ON 1 = 0;
 Unable to automatically determine a timestamp for your query; this can happen if your query depends on non-materialized sources
+
+# Test joins on implicitly coercible data types
```
Comment (justinj) [inline test/testdrive/joins.td:243]: Is there a reason this needs to be tested in testdrive as opposed to the sqllogictests?

File: src/sql/src/plan/query.rs [hunk 2]
```diff
@@ -1764,72 +1767,96 @@ fn plan_using_constraint(
...
+                .collect(),
+        ),
+        allow_aggregates: false,
+        allow_subqueries: false,
+    };
+
     for column_name in column_names {
-        let (l, _) = left_scope.resolve_column(column_name)?;
+        let (lhs, _) = left_scope.resolve_column(column_name)?;
         let (r, _) = right_scope.resolve_column(column_name)?;
```
Comment (justinj) [inline src/sql/src/plan/query.rs:1791]: nit: for consistency `s/r/rhs/`?

**Review by justinj — APPROVED** [review]
LGTM, do you have a source that the logic for coerce_homogenous_exprs is indeed what's needed here? Mostly just curious, seems right!

File: test/testdrive/joins.td (hunk 1 above)
Comment (benesch) [reply test/testdrive/joins.td:243]: Agreed that an SLT here would be best!

**Review by benesch — COMMENTED** [review]
Code looks good, but I think a small reordering (of stuff that existed before) could improve clarity. 1s!

**Comment by sploiselle** [comment]
@justinj really, really good q re: `coerce_homogeneous_exprs`––using this function lets us avoid going all the way through the function selection machinery (`func::resolve_func`, `func::select_impl` –– which is definitely an option) bc we know we can directly invoke the functions we want with homogeneously typed exprs. That we want to use these same homogeneously typed exprs twice then makes it really decisive that this is the better choice.

As for a source, the logic is that for the join to work, the two expressions must be of the same type, as evidenced by the fact that `BinaryFunc::Eq` and `VariadicFunc::Coalesce` both require homogeneously typed exprs (though I realize that this makes my argument circular ¯\\_(ツ)_/¯), and this is kind of implicit cast is allowed bc PG does it.

Not sure if I under-, over-, or incorrect-dimensionally-shot your question, but please lmk!

**Comment by benesch** [comment]
Ok, what do you think about restructuring the body of the loop in `plan_using_constraint` so it looks like this:

```rust
        let (lhs, _) = left_scope.resolve_column(column_name)?;
        let (mut rhs, _) = right_scope.resolve_column(column_name)?;
        if lhs.level != 0 || rhs.level != 0 {
            bail!(
                "Internal error: name {} in USING resolved to outer column",
                column_name
            )
        }

        // The new column can be named using aliases from either the right or
        // left.
        let mut names = left_scope.items[lhs.column].names.clone();
        names.extend(right_scope.items[rhs.column].names.clone());

        // Adjust the RHS reference to its post-join location.
        rhs.column += left_scope.len();

        // Join keys must be resolved to same type.
        let mut exprs = coerce_homogeneous_exprs(
            &format!("NATURAL/USING join column {}", column_name),
            &ecx,
            vec![
                CoercibleScalarExpr::Coerced(ScalarExpr::Column(lhs)),
                CoercibleScalarExpr::Coerced(ScalarExpr::Column(rhs)),
            ],
            None,
        )?;
        let (expr1, expr2) = (exprs.remove(0), exprs.remove(0));

        join_exprs.push(ScalarExpr::CallBinary {
            func: BinaryFunc::Eq,
            expr1: Box::new(expr1.clone()),
            expr2: Box::new(expr2.clone()),
        });
        map_exprs.push(ScalarExpr::CallVariadic {
            func: VariadicFunc::Coalesce,
            exprs: vec![expr1, expr2],
        });
        new_items.push(ScopeItem {
            names,
            expr: None,
            nameable: true,
        });
        dropped_columns.insert(lhs.column);
        dropped_columns.insert(rhs.column);
```

The idea is to do the thing that requires the unadjusted RHS column, adjust the RHS column, and then do all the things that require the adjusted RHS column.

**Comment by sploiselle** [comment]
@benesch Nice 🙇🙇🙇

**Review by benesch — APPROVED** [review]
Looks swell to me!

**Pull request merged**